Dockerfile

```diff
@@ -1,64 +1,58 @@
-FROM ubuntu:18.04 as builder
+ARG DEBIAN_VERSION=bullseye-slim
+ARG GO_VERSION=1.15-buster
+
+FROM golang:${GO_VERSION} AS builder
 
-# Install GCC
 RUN apt-get update && \
     apt-get install -y \
     gcc \
     cmake \
     wget \
-    curl \
-    npm 
-
-# Install node source for polkadotjs tests
-RUN curl -sL https://deb.nodesource.com/setup_14.x | bash -
-
-# Install nodejs for polkadotjs tests
-RUN apt-get update && \
-    apt-get install -y \
+    npm \
+    # Install nodejs for polkadotjs tests
     nodejs
 
-# Install Go
-RUN wget https://dl.google.com/go/go1.15.5.linux-amd64.tar.gz
-RUN tar -C /usr/local -xzf go1.15.5.linux-amd64.tar.gz
+# Install node source for polkadotjs tests
+RUN wget -qO- https://deb.nodesource.com/setup_14.x | bash -
 
 # Install subkey
-RUN wget -P /usr/local/bin/ https://chainbridge.ams3.digitaloceanspaces.com/subkey-v2.0.0
-RUN mv /usr/local/bin/subkey-v2.0.0 /usr/local/bin/subkey
-RUN chmod +x /usr/local/bin/subkey
+RUN wget -O /usr/local/bin/subkey https://chainbridge.ams3.digitaloceanspaces.com/subkey-v2.0.0 && \
+    chmod +x /usr/local/bin/subkey
 
-# Configure go env vars
-ENV GO111MODULE=on
-ENV GOPATH=/gocode
-ENV GOROOT=/usr/local/go
-ENV PATH=$PATH:$GOPATH/bin:$GOROOT/bin
+# Polkadot JS dependencies
+WORKDIR /go/src/github.com/ChainSafe/gossamer/tests/polkadotjs_test
+COPY tests/polkadotjs_test/package.json tests/polkadotjs_test/package-lock.json ./
+RUN npm install
 
-# Prepare structure and change dir
-RUN mkdir -p $GOPATH/src/github.com/ChainSafe/gossamer
-WORKDIR $GOPATH/src/github.com/ChainSafe/gossamer
+WORKDIR /go/src/github.com/ChainSafe/gossamer
 
-# Add go mod lock files and gossamer default config
-COPY go.mod .
-COPY go.sum .
-
-# Get go mods
+# Go dependencies
+COPY go.mod go.sum ./
 RUN go mod download
 
-# Copy gossamer sources
-COPY . $GOPATH/src/github.com/ChainSafe/gossamer
+# Prepare libwasmer.so for COPY
+RUN cp /go/pkg/mod/github.com/wasmerio/go-ext-wasm@*/wasmer/libwasmer.so libwasmer.so
 
-# Install js dependencies for polkadot.js tests
-RUN cd $GOPATH/src/github.com/ChainSafe/gossamer/tests/polkadotjs_test && npm install
+# Copy gossamer sources
+COPY . .
 
 # Build
 RUN GOBIN=$GOPATH/src/github.com/ChainSafe/gossamer/bin go run scripts/ci.go install
 
-# Create symlink
-RUN ln -s $GOPATH/src/github.com/ChainSafe/gossamer/bin/gossamer /usr/local/gossamer
+# Final stage based on Debian Bullseye Slim
+FROM debian:${DEBIAN_VERSION}
 
-# Give permissions
-RUN chmod +x $GOPATH/src/github.com/ChainSafe/gossamer/scripts/docker-entrypoint.sh
+# Install libwasmer.so
+ENV LD_LIBRARY_PATH=/lib:/usr/lib
+COPY --from=builder /go/src/github.com/ChainSafe/gossamer/libwasmer.so /lib/libwasmer.so
+
+EXPOSE 7001 8546 8540
 
-# Expose gossamer command and port
 ENTRYPOINT ["/gocode/src/github.com/ChainSafe/gossamer/scripts/docker-entrypoint.sh"]
 CMD ["/usr/local/gossamer"]
```

Comment:
I guess this `CMD` could be moved to `scripts/docker-entrypoint.sh`

Reply:
Well we could change a lot of changes regarding the entrypoint 😉

But I wanted to keep compatibility (and also cmd and entrypoint come as documentation when you `docker inspect chainsafe/gossamer`). If we're fine doing small compatibility breakings, then yes we can change that one. We could also have it in another shorter directory path than this one.

On a side note, I think (80% sure) relative paths for entrypoints do not work for example on `scratch` based Docker images, so we may want to keep using absolute paths. But that's just nitpicking from me 😄